taichi/ui/utils/utils.h

```diff
@@ -0,0 +1,218 @@
+
+
+#pragma once
+
+#include <string>
+#include <iostream>
+#include <stdexcept>
+#include <stdio.h>
+#include <iostream>
+#include <fstream>
+#include <stdexcept>
+#include <algorithm>
+#include <chrono>
+#include <vector>
+#include <cstring>
+#include <cstdlib>
+#include <cstdint>
+#include <array>
+#include <optional>
+#include <set>
+#include <unordered_map>
+
+#ifdef _WIN64
+#include <aclapi.h>
+#include <dxgi1_2.h>
+#include <windows.h>
+#include <VersionHelpers.h>
+#define _USE_MATH_DEFINES
+#endif
+
+#ifdef _WIN64
+#define VK_USE_PLATFORM_WIN32_KHR 1
+#endif
+#include <volk.h>
+#include <GLFW/glfw3.h>
+
+#include <stdarg.h>
+
+#define GLM_FORCE_DEPTH_ZERO_TO_ONE
+#include <glm/glm.hpp>
+#include <glm/gtc/matrix_transform.hpp>
+#include <glm/gtc/type_ptr.hpp>
+
+#define TI_UI_NAMESPACE_BEGIN \
+  namespace taichi {          \
+  namespace ui {
+
+#define TI_UI_NAMESPACE_END \
+  }                         \
+  }
+
+TI_UI_NAMESPACE_BEGIN
+
+inline void initGLFW() {
+  if (!glfwInit()) {
+    printf("cannot initialize GLFW\n");
+    exit(EXIT_FAILURE);
+  }
+}
+
+inline GLFWwindow *create_glfw_window_(const std::string &name,
+                                       int screenWidth,
+                                       int screenHeight,
+                                       bool vsync) {
+  initGLFW();
+  GLFWwindow *window;
+
+  glfwWindowHint(GLFW_CLIENT_API, GLFW_NO_API);
+
+  window = glfwCreateWindow(screenWidth, screenHeight, name.c_str(), nullptr,
+                            nullptr);
+
+  if (!window) {
+    glfwTerminate();
+    exit(EXIT_FAILURE);
+  }
+
+  if (vsync) {
+    glfwSwapInterval(1);
+  } else {
+    glfwSwapInterval(0);
```

Comment:
This does nothing in Vulkan, since this is handled by the swapchain. When we set present mode to FIFO, it will be VSYNC, if present mode is set to immediate or mailbox, it will not be vsync but it still needs to acquire images so some delays will be there.

Reply:
VSYNC stuff in vulkan will be taken care of properly. dont worry.

Reply:
This is here because we might still have a OpenGL-backed GGUI in the future.